bitcoin/src/musig2.rs

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Musig2
+//!
+//! Support for Musig2.
+//! 
+//! Musig2 is a Schnorr-based multi-signature scheme that is designed to be secure and efficient.
+//! It is a two-round protocol that is designed to be secure against rogue-key attacks.
+use core::fmt;
+
+use secp256k1::{musig::{new_musig_nonce_pair, MusigAggNonce, MusigKeyAggCache, MusigPubNonce, MusigSecNonce, MusigSession, MusigSessionId, MusigTweakErr}, Message, PublicKey, Scalar, Secp256k1, SecretKey, Signing, Verification};
+
+use crate::{key::TweakedPublicKey, ScriptBuf, XOnlyPublicKey};
+
+use crate::address::script_pubkey::ScriptBufExt;
+
+#[cfg(feature = "rand-std")]
+pub use secp256k1::rand;
+
+/// A Musig2 error
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum Error {
+    /// Invalid tweak operation
+    TweakErr,
+    /// Failed to generate nonce
+    NonceGenError,
+    /// Public nonce missing from the list of nonces
+    PubNonceMissing,
+    /// Null session
+    NullSession,
+    /// Partial signature error
+    PartialSignatureError,
+}
+
+internals::impl_from_infallible!(Error);
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::TweakErr => write!(f, "Invalid tweak operation"),
+            Error::NonceGenError => write!(f, "Failed to generate nonce. Supplied a zero session id."),
+            Error::PubNonceMissing => write!(f, "Public nonce missing from the list of nonces."),
+            Error::NullSession => write!(f, "Null session. Execute MusigData::process_nonces()."),
+            Error::PartialSignatureError => write!(f, "Partial signature error."),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for Error { }
+
+/// Creates a new [`secp256k1::musig::MusigKeyAggCache`] by supplying a list of PublicKeys used in the session.
+pub fn create_musig_key_agg_cache<C: Verification>(secp: &Secp256k1<C>, pubkeys: &[PublicKey]) -> MusigKeyAggCache {
+
+    MusigKeyAggCache::new(&secp, pubkeys)
+}
+
+/// Apply "x-only" tweaking to a public key in a [`secp256k1::musig::MusigKeyAggCache`].
+pub fn tweak_taproot_key<C: Verification>(secp: &Secp256k1<C>, musig_key_agg_cache: &mut MusigKeyAggCache, tweak: &Scalar) -> Result<PublicKey, Error>
+{
+    match musig_key_agg_cache.pubkey_xonly_tweak_add(secp, tweak) {
+        Ok(tweaked_key) => Ok(tweaked_key),
+        Err(MusigTweakErr::InvalidTweak) => Err(Error::TweakErr),
+    }
+}
+
+/// Create a new [`ScriptBuf`] for a P2TR output with a tweaked x-only public key.
+pub fn new_p2tr_script_buf(output_key: XOnlyPublicKey) -> ScriptBuf {
+    let tweaked_public_key = TweakedPublicKey::dangerous_assume_tweaked(output_key);
+    ScriptBuf::new_p2tr_tweaked(tweaked_public_key)
+}
```

Comment:
This function bypasses our API protections for no obvious reason.

Reply:
The tweak in this Musig2 implementation is done by `MusigKeyAggCache::pubkey_xonly_tweak_add()`, not by `UntweakedPublicKey::tap_tweak`.
So `ScriptBuf::new_p2tr_tweaked` is used to create the P2TR `ScriptBuf`.

I'm not sure which API protection you are referring to or what your suggestion would be for creating `ScriptBuf` from a tweaked key.

Reply:
Got it. The problem is removing `dangerous_assume_tweaked` from the name.
Anyway, I removed the `musig` file in this PR as it wasn't adding a good abstraction to the `secp256k1::musig` API.